src/translator/quality_estimator.cpp

```diff
@@ -0,0 +1,302 @@
+  size_t sentenceIndex = 0;
+
+  for (const auto& history : histories) {
+    const auto logProbs = std::get<1>(history->top())->tracebackWordScores();
+    response.qualityScores.push_back(computeSentenceScores(logProbs, response.target, sentenceIndex));
+
+    ++sentenceIndex;
+  }
+}
+
+Response::WordsQualityEstimate LogisticRegressorQualityEstimator::computeSentenceScores(
+    const std::vector<float>& logProbs, const AnnotatedText& target, const size_t sentenceIdx) const
+
+{
+  // @TODO: comment code below
+  const auto [subwordByWordBR, wordslogProbs] = remapWordsAndLogProbs(logProbs, target, sentenceIdx);
+
+  const auto wordQualityScores = predict(extractFeatures(wordslogProbs));
+
+  const float sentenceScore = std::accumulate(std::begin(wordQualityScores), std::end(wordQualityScores), float(0.0)) /
+                              wordQualityScores.size();
+
+  return {wordQualityScores, subwordToWords(subwordByWordBR), sentenceScore};
+}
+
+std::vector<float> LogisticRegressorQualityEstimator::predict(const Matrix& features) const {
+  std::vector<float> scores(features.rows);
+
+  for (int i = 0; i < features.rows; ++i) {
+    for (int j = 0; j < features.cols; ++j) {
+      scores[i] += features.at(i, j) * coefficientsByStds_[j];
+    }
+  }
+
+  /// Applies the linear model followed by a sigmoid function to each element
+
+  for (int i = 0; i < features.rows; ++i) {
+    scores[i] = std::log(1 - (1 / (1 + std::exp(-(scores[i] - constantFactor_ + intercept_)))));
+  }
+
+  return scores;
+}
+
+Matrix LogisticRegressorQualityEstimator::extractFeatures(const std::vector<std::vector<float>>& wordsLogProbs) {
+  if (wordsLogProbs.empty()) {
+    return std::move(Matrix(0, 0));
+  }
+
+  Matrix features(wordsLogProbs.size(), /*numFeatures =*/4);
+  size_t featureRow = 0;
+  const size_t I_MEAN{0}, I_MIN{1}, I_NUM_SUBWORDS{2}, I_OVERALL_MEAN{3};
+
+  float overallMean = 0.0;
+  size_t numlogProbs = 0;
+
+  for (const auto& wordLogProbs : wordsLogProbs) {
+    if (wordLogProbs.size() == 0) {
+      ++featureRow;
+      continue;
+    }
+
+    float minScore = std::numeric_limits<float>::max();
+
+    for (const auto& logProb : wordLogProbs) {
+      ++numlogProbs;
+      overallMean += logProb;
+      features.at(featureRow, I_MEAN) += logProb;
+
+      if (logProb < minScore) {
+        minScore = logProb;
+      }
+    }
+
+    features.at(featureRow, I_MEAN) /= static_cast<float>(wordLogProbs.size());
+    features.at(featureRow, I_MIN) = minScore;
+    features.at(featureRow, I_NUM_SUBWORDS) = wordLogProbs.size();
+
+    ++featureRow;
+  }
+
+  if (numlogProbs == 0) {
+    return std::move(Matrix(0, 0));
+  }
+
+  overallMean /= numlogProbs;
+
+  for (int i = 0; i < features.rows; ++i) {
+    features.at(i, I_OVERALL_MEAN) = overallMean;
+  }
+
+  return std::move(features);
+}
+
+std::pair<std::vector<std::vector<ByteRange>>, std::vector<std::vector<float>>> remapWordsAndLogProbs(
```

Comment:
We followed your suggestion. We refactored the method to return a vector of SubwordRange( as an alias to ByteRange) and because we are mapping when the words start and ends, we renamed it to mapWords. Commit  9374aeb